File: aicq-crypto/src/handshake.ts
/**
 * Authenticated key-exchange handshake (a simplified Signal / Noise
 * inspired protocol).
 *
 * Overview
 * --------
 * 1. **Initiator** creates a `HandshakeRequest` with their identity public
 *    key and an ephemeral public key.
 *
 * 2. **Responder** receives the request, performs an ephemeral-ephemeral
 *    DH, derives a session key (with identity keys mixed in for
 *    authentication binding), and returns a `HandshakeResponse` with their
 *    own public keys and a proof (HMAC of the transcript).
 *
 * 3. **Initiator** receives the response, independently computes the same
 *    DH shared secret, derives the session key, and verifies the
 *    responder's proof.
 *
 * Key derivation
 * ---------------
 * The session key is derived from the ephemeral-ephemeral DH shared
 * secret (ee), with all four public keys mixed in as HKDF context for
 * authentication binding.  This provides forward secrecy while tying the
 * session to both parties' long-term identity keys.
 */

import type { KeyPair, HandshakeRequest, HandshakeResponse } from "./types.js";
import { computeSharedSecret, deriveSessionKey } from "./keyExchange.js";
import { nacl } from "./nacl.js";

/* ------------------------------------------------------------------ */
/*  Transcript hashing helpers                                         */
/* ------------------------------------------------------------------ */

/**
 * Compute an HMAC-SHA-512 proof over the handshake transcript.
 */
function computeProof(
  key: Uint8Array,
  ...parts: Uint8Array[]
): Uint8Array {
  const transcript = new Uint8Array(
    parts.reduce((sum, p) => sum + p.length, 0),
  );
  let off = 0;
  for (const p of parts) {
    transcript.set(p, off);
    off += p.length;
  }

  // Simple HMAC: H(key ^ opad || H(key ^ ipad || message))
  const BLOCK = 128;
  const k = new Uint8Array(BLOCK);
  k.set(key.length > BLOCK ? nacl.hash(key) : key, 0);

  const iKey = new Uint8Array(BLOCK);
  const oKey = new Uint8Array(BLOCK);
  for (let i = 0; i < BLOCK; i++) {
    iKey[i] = k[i] ^ 0x36;
    oKey[i] = k[i] ^ 0x5c;
  }

  const inner = new Uint8Array(BLOCK + transcript.length);
  inner.set(iKey);
  inner.set(transcript, BLOCK);
  const innerHash = nacl.hash(inner);

  const outer = new Uint8Array(BLOCK + 64);
  outer.set(oKey);
  outer.set(innerHash, BLOCK);

  return nacl.hash(outer);
}

/**
 * Derive the shared session key from the ephemeral-ephemeral DH shared
 * secret and all four public keys (identity + ephemeral) for authentication
 * binding.
 */
function deriveHandshakeKey(
  ee: Uint8Array,
  initiatorIdPk: Uint8Array,
  initiatorEphPk: Uint8Array,
  responderIdPk: Uint8Array,
  responderEphPk: Uint8Array,
): Uint8Array {
  // Build a context string from all public keys for HKDF domain separation
  const contextLen = 32 + initiatorIdPk.length + initiatorEphPk.length +
                     responderIdPk.length + responderEphPk.length;
  const context = new Uint8Array(contextLen);
  let off = 0;

  // Use SHA-512 hash of the concatenated public keys as additional input
  const allPks = new Uint8Array(
    initiatorIdPk.length + initiatorEphPk.length +
    responderIdPk.length + responderEphPk.length,
  );
  allPks.set(initiatorIdPk, 0);
  allPks.set(initiatorEphPk, initiatorIdPk.length);
  allPks.set(responderIdPk, initiatorIdPk.length + initiatorEphPk.length);
  allPks.set(responderEphPk, initiatorIdPk.length + initiatorEphPk.length + responderIdPk.length);

  const pkHash = nacl.hash(allPks);
  const combined = new Uint8Array(32 + 64); // ee(32) + hash(64)
  combined.set(ee, 0);
  combined.set(pkHash, 32);

  return deriveSessionKey(combined, "aicq-handshake");
}

/* ------------------------------------------------------------------ */
/*  Create Handshake Request (Initiator)                               */
/* ------------------------------------------------------------------ */

/**
 * Create a handshake request to start an authenticated key exchange.
 *
 * The caller should generate a fresh ephemeral key pair before calling
 * this function.
 */
export function createHandshakeRequest(
  myPubKey: Uint8Array,
  myEphemeralPubKey: Uint8Array,
): HandshakeRequest {
  return {
    identityPublicKey: myPubKey,
    ephemeralPublicKey: myEphemeralPubKey,
  };
}

/* ------------------------------------------------------------------ */
/*  Create Handshake Response (Responder)                              */
/* ------------------------------------------------------------------ */

/**
 * Create a response to a handshake request.
 *
 * Computes the ephemeral-ephemeral DH shared secret and derives the
 * session key with all four public keys mixed in for authentication.
 * Produces a proof (HMAC of the transcript) to authenticate the responder.
 */
export function createHandshakeResponse(
  request: HandshakeRequest,
  myIdentityKeys: KeyPair,
  myEphemeralKeys: KeyPair,
): HandshakeResponse {
  // ee: ephemeral-ephemeral  (R_e × I_e)
  const ee = computeSharedSecret(
    myEphemeralKeys.secretKey,
    request.ephemeralPublicKey,
  );

  const sessionKey = deriveHandshakeKey(
    ee,
    request.identityPublicKey,      // initiator identity pk
    request.ephemeralPublicKey,    // initiator ephemeral pk
    myIdentityKeys.publicKey,      // responder identity pk
    myEphemeralKeys.publicKey,     // responder ephemeral pk
  );

  // Proof = HMAC over the entire handshake transcript
  const proof = computeProof(
    sessionKey,
    request.identityPublicKey,
    request.ephemeralPublicKey,
    myIdentityKeys.publicKey,
    myEphemeralKeys.publicKey,
  );

  return {
    identityPublicKey: myIdentityKeys.publicKey,
    ephemeralPublicKey: myEphemeralKeys.publicKey,
    proof,
  };
}

/* ------------------------------------------------------------------ */
/*  Complete Handshake                                                */
/* ------------------------------------------------------------------ */

/**
 * Complete the handshake and derive the shared session key.
 *
 * Works for **both** the initiator and the responder.  The function
 * detects which role the caller plays by comparing `myIdentityKeys.publicKey`
 * against `request.identityPublicKey` (initiator) or
 * `response.identityPublicKey` (responder).
 *
 * Both sides compute the same ephemeral-ephemeral DH and derive the
 * session key with the same set of public keys, so they arrive at the
 * same result.
 *
 * @returns The derived 32-byte session key.
 * @throws Error if the responder's proof does not validate.
 */
export function completeHandshake(
  response: HandshakeResponse,
  request: HandshakeRequest,
  myIdentityKeys: KeyPair,
  myEphemeralKeys: KeyPair,
): Uint8Array {
  // Determine which role we play
  const isInitiator = buffersEqual(myIdentityKeys.publicKey, request.identityPublicKey);

  let ee: Uint8Array;

  if (isInitiator) {
    // ee: my-ephemeral × their-ephemeral  (I_e × R_e)
    ee = computeSharedSecret(myEphemeralKeys.secretKey, response.ephemeralPublicKey);
  } else {
    // ee: my-ephemeral × their-ephemeral  (R_e × I_e, same as I_e × R_e)
    ee = computeSharedSecret(myEphemeralKeys.secretKey, request.ephemeralPublicKey);
  }

  const sessionKey = deriveHandshakeKey(
    ee,
    request.identityPublicKey,      // initiator identity pk
    request.ephemeralPublicKey,    // initiator ephemeral pk
    response.identityPublicKey,     // responder identity pk
    response.ephemeralPublicKey,    // responder ephemeral pk
  );

  // Recompute and verify the responder's proof
  const expectedProof = computeProof(
    sessionKey,
    request.identityPublicKey,
    request.ephemeralPublicKey,
    response.identityPublicKey,
    response.ephemeralPublicKey,
  );

  if (!buffersEqual(expectedProof, response.proof)) {
    throw new Error("Handshake proof verification failed");
  }

  return sessionKey;
}

/* ------------------------------------------------------------------ */
/*  Helpers                                                            */
/* ------------------------------------------------------------------ */

function buffersEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

File: shared/crypto/handshake.py
"""
Noise-XK-inspired authenticated key exchange handshake.

Implements a three-message handshake pattern:
1. Initiator → Responder: HandshakeRequest(identity_pub, ephemeral_pub)
2. Responder → Initiator: HandshakeResponse(identity_pub, ephemeral_pub, proof)

Key derivation:
1. Compute ee = X25519(ephemeral_secret, their_ephemeral_public)
2. Derive handshake key = HMAC-SHA512(ee, all_4_public_keys_concatenated)
3. Session key = first 32 bytes of the derived key
4. Proof = HMAC-SHA512(session_key, transcript_of_all_public_keys)

The transcript is the concatenation of all four public keys in canonical
order: initiator_identity + initiator_ephemeral + responder_identity +
responder_ephemeral.
"""

from typing import Tuple

from .types import KeyPair, HandshakeRequest, HandshakeResponse
from .key_exchange import compute_shared_secret, hmac_sha512


def _build_transcript(
    initiator_identity_pub: bytes,
    initiator_ephemeral_pub: bytes,
    responder_identity_pub: bytes,
    responder_ephemeral_pub: bytes,
) -> bytes:
    """Build the handshake transcript from all four public keys.

    The canonical ordering is: initiator_identity, initiator_ephemeral,
    responder_identity, responder_ephemeral.

    Args:
        initiator_identity_pub: Initiator's 32-byte identity public key.
        initiator_ephemeral_pub: Initiator's 32-byte ephemeral public key.
        responder_identity_pub: Responder's 32-byte identity public key.
        responder_ephemeral_pub: Responder's 32-byte ephemeral public key.

    Returns:
        The 128-byte concatenated transcript.
    """
    return (
        initiator_identity_pub
        + initiator_ephemeral_pub
        + responder_identity_pub
        + responder_ephemeral_pub
    )


def create_handshake_request(
    identity_pub_key: bytes,
    ephemeral_pub_key: bytes,
) -> HandshakeRequest:
    """Create a handshake request as the initiator.

    Args:
        identity_pub_key: The initiator's long-term Ed25519 public key (32 bytes).
        ephemeral_pub_key: The initiator's ephemeral X25519 public key (32 bytes).

    Returns:
        A HandshakeRequest containing both public keys.

    Raises:
        ValueError: If key lengths are incorrect.
    """
    if len(identity_pub_key) != 32:
        raise ValueError(
            f"Identity public key must be 32 bytes, got {len(identity_pub_key)}"
        )
    if len(ephemeral_pub_key) != 32:
        raise ValueError(
            f"Ephemeral public key must be 32 bytes, got {len(ephemeral_pub_key)}"
        )

    return HandshakeRequest(
        identity_public_key=identity_pub_key,
        ephemeral_public_key=ephemeral_pub_key,
    )


def create_handshake_response(
    request: HandshakeRequest,
    my_identity_keys: KeyPair,
    my_ephemeral_keys: KeyPair,
) -> Tuple[HandshakeResponse, bytes]:
    """Create a handshake response as the responder.

    Computes the shared secret, derives the session key, and generates
    an HMAC proof that binds all public keys to the session.

    Args:
        request: The initiator's HandshakeRequest.
        my_identity_keys: The responder's long-term identity KeyPair
            (Ed25519, public key used for transcript).
        my_ephemeral_keys: The responder's ephemeral KeyPair (X25519)
            used for the DH key exchange.

    Returns:
        A tuple of (HandshakeResponse, session_key) where:
        - HandshakeResponse contains the responder's public keys and proof.
        - session_key is the 32-byte derived session key.

    Raises:
        ValueError: If key lengths are incorrect or DH fails.
    """
    # Step 1: Compute ee = X25519(my_ephemeral_secret, their_ephemeral_public)
    ee = compute_shared_secret(
        my_ephemeral_keys.secret_key,
        request.ephemeral_public_key,
    )

    # Build transcript: initiator keys first, then responder keys
    transcript = _build_transcript(
        initiator_identity_pub=request.identity_public_key,
        initiator_ephemeral_pub=request.ephemeral_public_key,
        responder_identity_pub=my_identity_keys.public_key,
        responder_ephemeral_pub=my_ephemeral_keys.public_key,
    )

    # Step 2: Derive handshake key = HMAC-SHA512(ee, transcript)
    handshake_key = hmac_sha512(ee, transcript)

    # Step 3: Session key = first 32 bytes
    session_key = handshake_key[:32]

    # Step 4: Proof = HMAC-SHA512(session_key, transcript)
    proof = hmac_sha512(session_key, transcript)

    response = HandshakeResponse(
        identity_public_key=my_identity_keys.public_key,
        ephemeral_public_key=my_ephemeral_keys.public_key,
        proof=proof,
    )

    return response, session_key


def complete_handshake(
    response: HandshakeResponse,
    request: HandshakeRequest,
    my_identity_keys: KeyPair,
    my_ephemeral_keys: KeyPair,
) -> bytes:
    """Complete the handshake as the initiator.

    Computes the same shared secret and session key, then verifies
    the responder's proof to authenticate the handshake.

    Args:
        response: The responder's HandshakeResponse.
        request: The original HandshakeRequest that was sent.
        my_identity_keys: The initiator's long-term identity KeyPair
            (Ed25519, public key used for transcript).
        my_ephemeral_keys: The initiator's ephemeral KeyPair (X25519)
            used for the DH key exchange.

    Returns:
        The 32-byte session key.

    Raises:
        ValueError: If the proof verification fails, key lengths are
            incorrect, or the DH computation fails.
    """
    # Step 1: Compute ee = X25519(my_ephemeral_secret, their_ephemeral_public)
    ee = compute_shared_secret(
        my_ephemeral_keys.secret_key,
        response.ephemeral_public_key,
    )

    # Build transcript: initiator keys first, then responder keys
    transcript = _build_transcript(
        initiator_identity_pub=my_identity_keys.public_key,
        initiator_ephemeral_pub=my_ephemeral_keys.public_key,
        responder_identity_pub=response.identity_public_key,
        responder_ephemeral_pub=response.ephemeral_public_key,
    )

    # Step 2: Derive handshake key = HMAC-SHA512(ee, transcript)
    handshake_key = hmac_sha512(ee, transcript)

    # Step 3: Session key = first 32 bytes
    session_key = handshake_key[:32]

    # Step 4: Verify proof
    expected_proof = hmac_sha512(session_key, transcript)
    if not _constant_time_compare(response.proof, expected_proof):
        raise ValueError("Handshake proof verification failed")

    return session_key


def _constant_time_compare(a: bytes, b: bytes) -> bool:
    """Constant-time comparison of two byte strings.

    Uses hmac.compare_digest for timing-attack resistance.

    Args:
        a: First byte string.
        b: Second byte string.

    Returns:
        True if the strings are equal, False otherwise.
    """
    import hmac as _hmac
    return _hmac.compare_digest(a, b)
